frontend/app/controllers/concerns/spree/checkout/address_book.rb

```diff
@@ -32,21 +32,23 @@ def set_addresses
       end
 
       def normalize_addresses
-        return unless params[:state] == 'address' && @order.bill_address_id && @order.ship_address_id
+        return unless params[:state] == 'address' && @order.bill_address_id
 
-        # ensure that there is no validation errors and addresses were saved
-        return unless @order.bill_address && @order.ship_address
+        if @order.checkout_steps.include?('delivery')
+          return unless @order.ship_address_id
+        end
 
         bill_address = @order.bill_address
         ship_address = @order.ship_address
+
+        if params[:save_user_address].present?
```

Comment:
```suggestion
        if params[:save_user_address].present? && try_spree_current_user.present?
```

we would like to avoid unnecessary SQL calls if possible

Reply:
@damianlegawiec done.